Track line and column of tokens in tokenizer

AmaranthineCodices opened the pull request:
Wanted to have this here, since it's probably going to be important for sane parser errors! This will track the line and column that each token *starts* on in `usize` fields. There is a test case for this behavior.

~~This code works under the assumption that tokens are a single line. **This will break for multi-line string literals**, should we choose to parse them in the tokenizer.~~

This will cause issues with column counts when the source code contains UTF-8 characters - we'd need another crate to handle that (specifically getting the length, in grapheme clusters, of a UTF-8 string).

AmaranthineCodices committed:
track line and column of tokens in tokenizer

AmaranthineCodices committed:
add note about multi-line tokens

LPGhatguy commented on src/tokenizer.rs:
```diff
@@ -96,13 +114,20 @@ pub fn tokenize<'a>(source: &'a str) -> Result<Vec<Token<'a>>, TokenizeError<'a>
+                // This will cause problems if tokens match more than one line,
+                // but that seems really really unlikely - that would mean
+                // allowing whitespace in the tokens!
+                current_column += eaten_str.len();
```
We could add a parameter to Token that indicates the number of lines the token spanned. I think that multi-line strings are important enough to get right for preservation!

AmaranthineCodices committed:
refactor line counting

- move to separate method
- properly handle newlines in eaten tokens
- unit test the calculation process

LPGhatguy commented on src/tokenizer.rs:
```diff
@@ -43,11 +44,12 @@ lazy_static! {
     static ref PATTERN_CLOSE_PAREN: Regex = Regex::new(r"^(\))").unwrap();
 
     static ref PATTERN_WHITESPACE: Regex = Regex::new(r"^\s+").unwrap();
+    static ref PATTERN_CHARS_AFTER_NEWLINE: Regex = Regex::new(r"\n([^\n]+)$").unwrap();
```
Does this regex need to be pinned to the beginning of the string? We don't need to get a bunch of captures from it, just one.

AmaranthineCodices replied:
It's not pinned to the beginning, it's pinned to the end, so that a string that eats multiple newlines doesn't get false-positives. The `^` character is in a character class - it's inverting the class, not pinning the pattern to the beginning of the string.

LPGhatguy replied:
Oh, no, I was referring to the lack of a `^` at the beginning of the string, which *would* pin to the beginning.

I don't completely understand why it isn't pinned at both ends?

AmaranthineCodices replied:
Because if the eaten string contains multiple newlines, pinning this regex to the beginning of the string wouldn't work. The regex could be written as something like `^.*\n([^\n]+)$` but that seems more than a little error-prone - it was easier to write it in terms of only the end of the string.

LPGhatguy replied:
Okay, I see now. I think I misunderstood the inputs and outputs at first.

LPGhatguy commented on src/tokenizer.rs:
```diff
@@ -43,11 +44,12 @@ lazy_static! {
 /// Tries to matches the given pattern against the string slice.
 /// If it does, the 'tokenizer' fn is invokved to turn the result into a token.
-fn try_advance<'a, F>(source: &'a str, pattern: &Regex, tokenizer: F) -> Option<(&'a str, TokenKind<'a>)>
+fn try_advance<'a, F>(source: &'a str, pattern: &Regex, tokenizer: F) -> Option<(&'a str, &'a str, TokenKind<'a>)>
```
This is probably at the point where I'd want to extract out a named struct for this function's return value iinstead of a tuple

AmaranthineCodices replied:
Yeah, that's probably for the best! I'll pull that out.

Edit: Should be good now!

AmaranthineCodices committed:
Pull try_advance results into a separate struct

LPGhatguy commented:
(Close/reopen to get Travis to build now)

LPGhatguy commented on src/tokenizer.rs:
```diff
@@ -72,17 +85,47 @@ fn eat<'a>(source: &'a str, pattern: &Regex) -> (&'a str, Option<&'a str>) {
+        }
+    }
+    // Otherwise we can just increment the current column by the length of the eaten chars
+    else {
```
I think that the Rust convention would be to format this block (and others like it, with leading comments) like:

```rs
} else {
    // Otherwise we can just increment the current column by the length of the eaten chars
}
```

AmaranthineCodices replied:
Okay, cool! Fixed now.

LPGhatguy commented on src/tokenizer.rs:
```diff
@@ -72,17 +85,47 @@ fn eat<'a>(source: &'a str, pattern: &Regex) -> (&'a str, Option<&'a str>) {
     let mut tokens = Vec::new();
     let mut current = source;
+    let mut current_line: usize = 1;
+    let mut current_column: usize = 1;
```
Do we need these type annotations?

AmaranthineCodices replied:
Apparently not! I was under the impression Rust would assume they were `i32` or something, but apparently it correctly deduces the type.

AmaranthineCodices committed:
remove extraneous type annotations

AmaranthineCodices committed:
Move comments inside their blocks

LPGhatguy commented:
Thanks!
